src/server/pfs/server/driver.go

```diff
@@ -2443,6 +2523,58 @@ func nodeToFileInfo(commit *pfs.Commit, path string, node *hashtree.NodeProto, f
 	return fileInfo
 }
 
+// nodeToFileInfoInputCommit is like nodeToFileInfo, but handles the case (which
+// currently only occurs in input commits) where files have a header that is
+// stored in their parent directory
+func nodeToFileInfoInputCommit(commit *pfs.Commit, filePath string,
+	node *hashtree.NodeProto, tree hashtree.HashTree, full bool) (*pfs.FileInfo, error) {
+	if node.FileNode == nil || !node.FileNode.HasHeaderFooter {
+		return nodeToFileInfo(commit, filePath, node, full), nil
+	}
+	node = proto.Clone(node).(*hashtree.NodeProto)
+	// validate baseFileInfo for logic below--if input hashtrees start using
+	// blockrefs instead of objects, this logic will need to be adjusted
+	if node.FileNode.Objects == nil {
+		return nil, fmt.Errorf("input commit node uses blockrefs; cannot apply header")
+	}
+
+	// 'file' includes header from parent—construct synthetic file info that
+	// includes header in list of objects & hash
+	parentPath := path.Dir(filePath)
+	parentNode, err := tree.Get(parentPath)
```

Comment:
It seems like things might get easier if we moved `Shared` into the `FileNode`s rather than having to recurse back to the parent. Might be too much data bloat though.

Reply:
We could, though if we did that then changing a directory's headers after putting some initial data would mean updating the `Shared` field of every child, which we currently get to avoid

Reply:
Isn't it illegal to update though?

Reply:
Oh, so after our conversation yesterday, I made it legal to update the header. Basically, `InspectFile` et al just include compute file headers at read time, so if change a header you don't have to worry about updating the hashes or anything and pipelines and such work great. One of the tests I just added actually covers this case now

Reply:
Right, makes sense.